app/services/account_search_service.rb

```diff
@@ -1,26 +1,90 @@
 # frozen_string_literal: true
 
 class AccountSearchService < BaseService
+  attr_reader :query, :limit, :resolve, :account
+
   def call(query, limit, resolve = false, account = nil)
-    return [] if query.blank? || query.start_with?('#')
+    @query = query
+    @limit = limit
+    @resolve = resolve
+    @account = account
+
+    search_service_results
+  end
+
+  private
 
-    username, domain = query.gsub(/\A@/, '').split('@')
-    domain = nil if TagManager.instance.local_domain?(domain)
+  def search_service_results
+    return [] if query_blank_or_hashtag?
 
-    if domain.nil?
-      exact_match = Account.find_local(username)
-      results     = account.nil? ? Account.search_for(username, limit) : Account.advanced_search_for(username, account, limit)
+    if resolving_non_matching_remote_account?
+      [FollowRemoteAccountService.new.call("#{query_username}@#{query_domain}")]
     else
-      exact_match = Account.find_remote(username, domain)
-      results     = account.nil? ? Account.search_for("#{username} #{domain}", limit) : Account.advanced_search_for("#{username} #{domain}", account, limit)
+      search_results_and_exact_match.compact.uniq
     end
+  end
+
+  def resolving_non_matching_remote_account?
+    resolve && !exact_match && !domain_is_local?
+  end
+
+  def search_results_and_exact_match
+    search_results.to_a + [exact_match]
+  end
+
+  def query_blank_or_hashtag?
+    query.blank? || query.start_with?('#')
+  end
+
+  def split_query_string
+    @_split_query_string ||= query.gsub(/\A@/, '').split('@')
+  end
+
+  def query_username
+    @_query_username ||= split_query_string.first
+  end
+
+  def query_domain
+    @_query_domain ||= query_without_split? ? nil : split_query_string.last
+  end
+
+  def query_without_split?
+    split_query_string.size == 1
+  end
+
+  def domain_is_local?
+    @_domain_is_local ||= TagManager.instance.local_domain?(query_domain)
+  end
 
-    results = [exact_match] + results.reject { |a| a.id == exact_match.id } if exact_match
+  def exact_match
+    @_exact_match ||= if domain_is_local?
+      Account.find_local(query_username)
+    else
+      Account.find_remote(query_username, query_domain)
```

Comment:
If you're nullifying query_domain, this method can be simplified to just Account.find_remote, because find_local is just an alias for find_remote with nil domain